### PR TITLE
New control frames

### PR DESCRIPTION
**Pull request by rschell**
PR for Issue #1572.  

Here are two new control frames, both using slider controls to adjust the step size or step rate.   The first commit adds fixed step size sliders for the Z and XY axes.  The second commit adds a continuous control frame using a single slider to adjust the step rate for all three axes.  Either or both can replace the current control frame by making the appropriate selection in the bCNC startup files.

No pride of authorship here, take it if you like it, adjust it or I'm happy to make any suggested changes.

Ron

### CODE REVIEW

**Comment by blayno** [comment]
This is the idea i was looking for, how do iinstall your code ?

**Comment by rschell** [comment]
> 
> 
> This is the idea i was looking for, well done, how do i install your code ?

If you down load the code from my fork rschell/bCNC using the New_Control_Frames branch, you will be using the code that I have been running on my raspberry Pi based machine.

**Comment by blayno** [comment]
i downloaded your code as zip file, how do i get it to run ?, i installed regular bcnc using pip install bCNC and i open it by double clicking the bCNC.bat file. when i double click the bCNC.bat file in your new control frames master it just flashes the terminal at me but doesnt open, i tried the make-exe.bat file but it doesnt work properly i cannot connect to grbl, the only way ive gotten it to run is by opening the __main__.py file in sublime text and running it that way. how do i install it through pip so i can use the bCNC.bat file ?

**Comment by blayno** [comment]
i really like the features you have implemented, brilliant work

**Comment by bosd** [comment]
@rschell Thanks for your work. Could you attend to the codefactor issues?

**Comment by rschell** [comment]
Happy to, if there is a plan to merge the branch

**Comment by bosd** [comment]
Just simple GUI test here. Is it on purpose to show all new control frames as default?
On some screens we might run into insufficient screenspace.

image

File: bCNC/controllers/GRBL1.py [hunk 1]
```diff
@@ -8,6 +8,8 @@
 import time
 import Utils
 
+OV_JOG_CANCEL   = chr(0x85)
```
Comment (bosd) [inline bCNC/controllers/GRBL1.py:11]: Could you remove the extra spaces here. To silence codefactor.

**Comment by bosd** [comment]
Testing this code..
Pressing a button in continieus control results in:

```
Traceback (most recent call last):
  File "D:\Users\user\AppData\Local\Programs\Python\Python310\lib\site-packages\bCNC\Utils.py", line 526, in __call__
    return self.func(*args)
  File "D:\Users\user\AppData\Local\Programs\Python\Python310\lib\site-packages\bCNC\ControlPage.py", line 1985, in jogcancel
    self.app.mcontrol.cjogcancel()
  File "D:\Users\user\AppData\Local\Programs\Python\Python310\lib\site-packages\bCNC\controllers\GRBL1.py", line 49, in cjogcancel
    self.master.serial_write(OV_JOG_CANCEL)
  File "D:\Users\user\AppData\Local\Programs\Python\Python310\lib\site-packages\bCNC\Sender.py", line 479, in serial_write
    ret = self.serial.write(data.encode())
AttributeError: 'NoneType' object has no attribute 'write'
```

File: bCNC/controllers/GRBL1.py (hunk 1 above)
Comment (rschell) [reply bCNC/controllers/GRBL1.py:11]: Already fixed in my local version, need to push the change

**Comment by rschell** [comment]
> 

It appears you were not connected to the grbl controller, I get this error too when disconnected.  Will look into trapping this condition.  It is a more generic issue than the continuous control frame.

**Comment by Git-Bruno** [comment]
FYI, I had to change spaces to tabs in Sender.py lines 472-474 to get your fork running.

**Comment by rschell** [comment]
> FYI, I had to change spaces to tabs in Sender.py lines 472-474 to get your fork running.

sorry, dealing with two versions one with tabs and one without.  Editor defaults to inserting 4 spaces.

**Comment by bosd** [comment]
FYI the codefactor PR is merged now.
So the version with spaces can be used.

**Pull request closed without merging**